### PR TITLE
Implement the consumption step.

### PR DESCRIPTION
**Pull request by n1k0**
[User Story](https://www.notion.so/ETAPES-Introduire-l-tape-cuisson-chez-le-consommateur-dans-le-calculateur-d55f51c2aba44686aaf50d92c3b190d8)

### CODE REVIEW

**Comment by n1k0** [comment]
@magopian @ccomb @paulboosz juste au cas où, même si le rapport cru/cuit n'est pas encore géré on peut tout à fait mettre en prod ce patch et continuer d'itérer ensuite ;)

**Review by magopian — COMMENTED** [review]
Nice work! Quite a big diff, well done.

I don't see anything to limit the number of preparations selected from the API? Is that going to be added later?

The computation of the Preparation impact are not very clear to me, and don't seem to apply to the docs, I'm not sure what I'm missing.

File: data/food/export_agb/export_builder.py [hunk 1]
```diff
@@ -51,6 +51,13 @@
     "Steel, unalloyed {RER}| steel production, converter, unalloyed | Cut-off, S - Copied from Ecoinvent": "steel",
     "Polyvinylchloride, suspension polymerised {RER}| polyvinylchloride production, suspension polymerisation | Cut-off, S - Copied from Ecoinvent": "pvc",
     "Aluminium, primary, ingot {RoW}| production | Cut-off, S - Copied from Ecoinvent": "aluminium",
+    "Electricity, low voltage {FR}| market for | Cut-off, S - Copied from Ecoinvent": "electricity",
+    "Tap water {Europe without Switzerland}| market for | Cut-off, S - Copied from Ecoinvent": "tapwater",
+    "Transport, freight, aircraft {RER}| intercontinental | Cut-off, S - Copied from Ecoinvent":"plane",
+    "Transport, freight, sea, transoceanic ship {GLO}| processing | Cut-off, S - Copied from Ecoinvent": "boat",
```
[the remark this answers is not in the document]
Comment (n1k0) [reply data/food/export_agb/export_builder.py:57]: It's not me it's @paulboosz, anyway they're defined in WellKnown, and used in transport computation formulas 
image

**Comment by n1k0** [comment]
> The computation of the Preparation impact are not very clear to me, and don't seem to apply to the docs, I'm not sure what I'm missing.

Would be sweet to elaborate a little so I can do anything about it.

File: openapi.yaml [hunk 2]
```diff
@@ -768,6 +769,47 @@ components:
         frozen:
           summary: Congelé
           value: frozen
+    preparationParam:
+      name: preparation[]
+      in: query
+      description: |
+        Liste des techniques de préparation mobilisées **à l'étape de consommation** :
```
[the remark this answers is not in the document]
Comment (n1k0) [reply openapi.yaml:776]: Done

File: openapi.yaml [hunk 3]
```diff
@@ -768,6 +769,47 @@ components:
...
+
+        - `frying`: Friture
+        - `pan-cooking`: Cuisson à la poêle
+        - `pan-warming`: Réchauffage à la poêle
+        - `oven`: Four
+        - `microwaves`: Four micro-ondes
```
[the remark this answers is not in the document]
Comment (n1k0) [reply openapi.yaml]: Done

File: src/Data/Food/Builder/Recipe.elm [hunk 4]
```diff
@@ -440,8 +460,8 @@ encodeProcess p =
         ]
 
 
-encodeQuery : Query -> Encode.Value
-encodeQuery q =
+encode : Query -> Encode.Value
+encode q =
     Encode.object
         [ ( "ingredients", Encode.list encodeIngredient q.ingredients )
         , ( "transform", q.transform |> Maybe.map encodeProcess |> Maybe.withDefault Encode.null )
```
[the remark this answers is not in the document]
Comment (n1k0) [reply src/Data/Food/Builder/Recipe.elm:467]: Done

File: src/Data/Food/Builder/Query.elm [hunk 5]
```diff
@@ -138,6 +154,7 @@ carrotCake =
           }
         ]
     , distribution = Retail.ambient
+    , preparation = []
```
[the remark this answers is not in the document]
Comment (n1k0) [reply src/Data/Food/Builder/Query.elm]: Done

**Comment by n1k0** [comment]
@magopian I addressed all your remarks, would you mind giving the patch a new look?

**Review by magopian — APPROVED** [review]
This looks great, thanks for the changes!

Nit (for later?): as the ratio for elec and heat is actually the same (if elec's ratio is 60%, then the heat ratio is 40%), I guess we could represent the elec-heat data using a custom type to avoid representing impossible states, something like

```elm
type EnergyNeeded = ElecAndHeat Energy Energy Unit.Ratio
```

What do you think?

**Comment by n1k0** [comment]
> What do you think?

Not so sure about that as the ratios may evolve rather fast when users (GTalgo, brands, etc) provide feedback about that. If you really want improvements for this data structure, please file a card about it and flag it "tech-debt".

**Pull request merged**